src/ds/ring_buffer.h

```diff
@@ -144,8 +145,12 @@ namespace ringbuffer
   {
     static inline uint64_t read64_impl(const BufferDef& bd, size_t index)
     {
-      uint64_t r = *reinterpret_cast<volatile uint64_t*>(bd.data + index);
-      atomic_thread_fence(std::memory_order_acq_rel);
+      // __atomic_load is used instead of std::atomic_ref since it's not
```

Comment:
Could you `#ifdef` the code using `__cpp_lib_atomic_ref` as a C++ feature flag, so that the intended code is also written here?

Reply:
Sounds nice idea. My concern would be we can't test the code now. But even it has a problem, it can be detected as a CI error when libc++ is upgraded and atomic_ref. is added.

Reply:
Fixed [59e65a9](https://github.com/microsoft/CCF/pull/5135/commits/59e65a94d714150aa4b3928503e279dea28fc04c)

Reply:
Looks like atomic_ref's constructor requires reference rather than pointer.
https://en.cppreference.com/w/cpp/atomic/atomic_ref/atomic_ref

Piccolo project uses libstdc++ and failed to compile because of the above reason.
Hopefully [83a7e65](https://github.com/microsoft/CCF/pull/5135/commits/83a7e65d2418f76f7a56ca25f6c6799d8092f9c8) fixes the CI.

Reply:
To be clear, currently our build looks like this.

- CCF and tests: Uses libc++ -> `__cpp_lib_atomic_ref ` is not defined -> uses `__atomic_load`
- (maybe only) Piccolo: Uses libstdc++ because its feature is needed ->  `__cpp_lib_atomic_ref ` is defined -> uses `atomic_ref`

So there is tradeoff between having ideal implementation and having multiple target compiled differently.
Personally I am fine with both.

Reply:
Piccolo uses libstdc++ because some of the libraries it depends on (arrow, parquet), installed through apt, are built against it. I am surprised that it ends up including the ring buffer header though, it certainly doesn't need or use it. Regardless, it's nice to have the branch using atomic_ref, support will probably land in libc++ at some point.